hooks/terraform_validate.sh

```diff
@@ -43,16 +43,97 @@ function per_dir_hook_unique_part {
   local -a -r args=("$@")
 
   local exit_code
-  local validate_output
+  #
+  # Get hook settings
+  #
+  local retry_once_with_cleanup=false
+
+  IFS=";" read -r -a configs <<< "${HOOK_CONFIG[*]}"
+
+  for c in "${configs[@]}"; do
+
+    IFS="=" read -r -a config <<< "$c"
+    key=${config[0]}
+    value=${config[1]}
+
+    case $key in
+      --retry-once-with-cleanup)
+        retry_once_with_cleanup=$value
+        ;;
+    esac
+  done
 
   common::terraform_init 'terraform validate' "$dir_path" || {
     exit_code=$?
     return $exit_code
   }
 
-  # pass the arguments to hook
-  validate_output=$(terraform validate "${args[@]}" 2>&1)
-  exit_code=$?
+  function do_validate {
+    validate_output=$(terraform validate "${args[@]}" 2>&1)
+    exit_code=$?
+    return $exit_code
+  }
+
+  function parse_validate {
+    # Requires jq
+    local exit_code
+    local validate_output
+    local valid
+    local summary
+
+    validate_output=$(terraform validate -json "${args[@]}" 2>&1)
+    exit_code=$?
+
+    valid=$(jq -rc '.valid' <<< "$validate_output")
+
+    if [ "$valid" == "true" ]; then
+      return 0
+    fi
+
+    # Pretty-print error information
+    jq '.diagnostics[]' <<< "$validate_output"
+
+    # Parse error message, return code 10 to indicate a catch
+    while IFS= read -r error_message; do
+      summary=$(jq -rc '.summary' <<< "$error_message")
+      case $summary in
+        "missing or corrupted provider plugins")
+          return 10
+          ;;
+        "Module source has changed")
+          return 10
+          ;;
+        "Module version requirements have changed")
+          return 10
+          ;;
+        "Module not installed")
+          return 10
+          ;;
+      esac
+    done < <(jq -rc '.diagnostics[]' <<< "$validate_output")
+    # Return `terraform validate`'s original exit code
+    # when `$summary` isn't covered by `case` block above
+    return $exit_code
+  }
+
+  if [ "$retry_once_with_cleanup" == "true" ]; then
+    parse_validate
+    exit_code=$?
+  else
+    do_validate
+    exit_code=$?
+  fi
+
+  if [ $exit_code -eq 10 ] && [ "$retry_once_with_cleanup" == "true" ]; then
+    if [ -d .terraform ]; then
+      # Will only be displayed if validation fails again.
+      common::colorify "yellow" "Validation failed. Removing .terraform from: $dir_path"
+      rm -rf .terraform
```

Comment:
Let's address https://github.com/antonbabenko/pre-commit-terraform/issues/203#issuecomment-918791847 and make it from workaround to useful for all solutions (also need changes to README)
```suggestion
      common::colorify "yellow" "Validation failed. Removing cached providers and modules from: $dir_path/.terraform"
      # Remove all files in the `.terraform` dir except `environment` - where stored info about TF workspaces.
      rm -rf .terraform/!("environment")
```

Reply:
Not sure about default value of `extglob` across different versions of Bash and don't like to take a chance on it, so let's be more explicit on what we need to delete then: `rm -rf .terraform/{modules,providers}/` — this should be very much sufficient to `Removing cached providers and modules from` and needs no Bash-specific quirks.

Just for reference here's [the description](https://developer.hashicorp.com/terraform/cli/init#working-directory-contents) of `.terraform` dir:
> A hidden `.terraform` directory, which Terraform uses to manage cached provider plugins and modules, record which [workspace](https://developer.hashicorp.com/terraform/cli/workspaces) is currently active, and record the last known backend configuration in case it needs to migrate state on the next run. This directory is automatically managed by Terraform, and is created during initialization.

PS: I don't count in custom setups like https://developer.hashicorp.com/terraform/cli/config/config-file#explicit-installation-method-configuration 😄

Reply:
You're right
```suggestion
      common::colorify "yellow" "Validation failed. Removing cached providers and modules from: $dir_path/.terraform"
      # .terraform` dir includes could some extra files, like `environment`,
      # where stored info about TF workspaces, so we can't just remove
      # `.terraform` dir complitely.
      rm -rf .terraform/{modules,providers}/
```

Reply:
Not sure how to suggest change to your suggested change, so posting it as is:
```
      # `.terraform` dir may comprise some extra files, like `environment`
      # which stores info about current TF workspace, so we can't just remove
      # `.terraform` dir completely.
```